library/translation.inc.php

```diff
@@ -1,9 +1,124 @@
+
+        // Check for conflicting settings
+        if ($lang_id === 1 && $globals->getBoolean('translate_skip_english_lookup')) {
+            (new SystemLogger())->warning(
+                "Both 'translate_preload_cache' and 'translate_skip_english_lookup' are enabled. " .
+                "These settings are mutually exclusive for English. Using 'translate_skip_english_lookup' (skipping cache warm)."
+            );
+            return;
+        }
+
+        $cache = &xlGetCache();
+        if (isset($cache[$lang_id])) {
+            // Skip if already warmed (or marked as empty) for this language
+            return;
+        }
+
+        // Load all translations for this language in a single query
+        $sql = <<<'SQL'
+            SELECT lang_constants.constant_name,
+                   lang_definitions.definition
+              FROM lang_definitions
+              JOIN lang_constants
+                ON lang_definitions.cons_id = lang_constants.cons_id
+             WHERE lang_definitions.lang_id = ?;
+        SQL;
+        $rows = QueryUtils::fetchRecordsNoLog($sql, [$lang_id]);
+        $cache[$lang_id] = array_column($rows, 'definition', 'constant_name');
+    }
+}
+
+/**
+ * Checks if there are any translations for the given language.
+ * Caches the result: empty array means "checked, no translations exist".
+ */
+function xlLangHasTranslations(int $lang_id): bool
+{
+    if ($lang_id === 1 && OEGlobalsBag::getInstance(true)->getBoolean('translate_skip_english_lookup')) {
+        return false;
+    }
+
+    $cache = &xlGetCache();
+
+    // Already have cached translations
+    if (!empty($cache[$lang_id])) {
+        return true;
+    }
+
+    // Already checked and found none
+    if (isset($cache[$lang_id])) {
+        return false;
+    }
+
+    // Check database for any translations
+    $sql = <<<'SQL'
+        SELECT 1
+          FROM lang_definitions
+          JOIN lang_constants
+            ON lang_definitions.cons_id = lang_constants.cons_id
+         WHERE lang_definitions.lang_id = ?
+         LIMIT 1;
+    SQL;
+    $rows = QueryUtils::fetchRecordsNoLog($sql, [$lang_id]);
+
+    if (empty($rows)) {
+        // Mark as "no translations" so future lookups short-circuit
+        $cache[$lang_id] = [];
+        return false;
+    }
+
+    return true;
+}
+
+/**
+ * Determine if we should skip translation and just return the input.
+ *
+ * When 'temp_skip_translations' is set
+ * When 'translate_skip_english_lookup' is set and the language is English
+ * When cache is empty (no translations)
+ * When first access to database finds there are no translations for that language.
+ */
+function xlskip(int $lang_id): bool
+{
+    $cache = &xlGetCache();
+    $globals = OEGlobalsBag::getInstance(true);
+    return ($globals->getBoolean('temp_skip_translations')
+        || ($lang_id === 1 && $globals->getBoolean('translate_skip_english_lookup'))
```

Comment:
Is there a constant we can reference for english instead of `=== 1` here? Even if it alone gets a special-cased addition I think it'd add clarity here.